privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,582 @@
+            * fbtoken
+            * pubkey
+
+        This endpoint is also used, if the smartphone sends the signed response
+        to the challenge during authentication
+        Parameters sent:
+            * serial
+            * nonce (which is the challenge)
+            * signature (which is the signed nonce)
+
+
+        :param request: The Flask request
+        :param g: The Flask global object g
+        :return: dictionary
+        """
+        details = {}
+        result = False
+        serial = getParam(request.all_data, "serial", optional=False)
+
+        if serial and "fbtoken" in request.all_data and "pubkey" in request.all_data:
+            # Do the 2nd step of the enrollment
+            try:
+                token_obj = get_one_token(serial=serial,
+                                          tokentype="push",
+                                          rollout_state="clientwait")
+                token_obj.update(request.all_data)
+            except ResourceNotFoundError:
+                raise ResourceNotFoundError("No token with this serial number in the rollout state 'clientwait'.")
+            init_detail_dict = request.all_data
+
+            details = token_obj.get_init_detail(init_detail_dict)
+            result = True
+        elif serial and "nonce" in request.all_data and "signature" in request.all_data:
+            challenge = getParam(request.all_data, "nonce")
+            serial = getParam(request.all_data, "serial")
+            signature = getParam(request.all_data, "signature")
+
+            # get the token_obj for the given serial:
+            token_obj = get_one_token(serial=serial, tokentype="push")
+            pubkey_pem = token_obj.get_tokeninfo(PUBLIC_KEY_SMARTPHONE)
+            if not pubkey_pem.startswith("-----"):
+                pubkey_pem = "-----BEGIN PUBLIC KEY-----\n{0!s}\n-----END PUBLIC KEY-----".format(pubkey_pem.strip().replace(" ", "+"))
+            # Do the 2nd step of the authentication
+            # Find valid challenges
+            challengeobject_list = get_challenges(serial=serial, challenge=challenge)
+
+            if challengeobject_list:
+                # There are valid challenges, so we check this signature
+                for chal in challengeobject_list:
+                    # verify the signature of the nonce
+                    pubkey_obj = serialization.load_pem_public_key(to_bytes(pubkey_pem), default_backend())
+                    sign_data = u"{0!s}|{1!s}".format(challenge, serial)
+                    try:
+                        pubkey_obj.verify(b32decode(signature),
+                                          sign_data.encode("utf8"),
+                                          padding.PKCS1v15(),
+                                          hashes.SHA256())
+                        # The signature was valid
+                        chal.set_otp_status(True)
+                        result = True
+                    except Exception as e:
```

Comment:
Hm sorry, one more thing: I guess it would be a good idea to only catch ``InvalidSignature`` here? We don't notice if anything else goes wrong otherwise (e.g. the signature was valid, but ``set_otp_status(True)`` fails).

Reply:
done in 0550f7f3